### ISSUE TITLE
To reinvent the processor - Veedrac

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4946]
To reinvent the processor - Veedrac<br>
<br>
https://ift.tt/2Zu5UNC<br>
<br>
Veedrac<br>
<br>
<div><div><section><div>
<div><div><div><div><a href="https://medium.com/@veedrac?source=post_page---------------------------"><img alt="Veedrac" width="48" src="https://miro.medium.com/fit/c/96/96/1*gQSmKyVTWg_GC8aI3ruRTg.png" height="48"></a></div></div></div></div>
<p><em>This is a very long, very dense, and very technical foray into CPU architecture. I’ve tried to make this approachable to enthusiast non-professionals, but if you don’t know roughly how CPUs work, you will struggle.</em></p>
<pre><span><strong>Introduction</strong></span><span><strong>The fundamental work of an instruction<br> — </strong>Coarse-Grained Out-of-Order (CG-OoO)<br> <strong>—</strong> The Mill</span><span><strong>Dependencies, not order</strong><br> <strong>1.</strong> False dependencies<br> <strong>2.</strong> Control Flow<br> <strong>1.</strong> Branch instructions impose ordering constraints ...<br> <strong>Idea:</strong> Branch Predict and Branch Verify<br> <strong>2.</strong> Branches outside the current context are opaque ...<br> <strong>3.</strong> Variable-latency instructions</span><span><strong>Memory</strong><br> <strong>— </strong>Baseline approaches<br> <strong>— </strong>Mill Computing's approach<br> <strong>Issue 1:</strong> Function boundaries still prevent ...<br> <strong>Issue 2:</strong> Loads potentially in the critical ...<br> <strong>Issue 3:</strong> Converging dataflow<br> <strong>Issue 4:</strong> It interacts badly with loop pipelining<br> <strong>— </strong>Mill Computing's approach, cont.<br> <strong>— </strong>Ideas for a better load<br> <strong>Idea 1:</strong> Prodigy-Mill hybrid approach<br> <strong>Idea 2:</strong> Restricted Asynchronous Dataflow<br> <strong>Idea 3:</strong> Source-agnostic argument pickups<br> <strong>Idea 4:</strong> Standard prefetch<br> <strong>— </strong>Security</span><span><strong>Concluding Remarks</strong></span></pre>
<p>I’m tired of the dominance of the out-of-order processor. They are large and wasteful, the ever-popular x86 is especially poor, and they are hard to understand. Their voodoo would be more appreciated if they pushed better at the limits of computation, but it’s obvious that the problems people solve have a latent inaccessible parallelism far in excess of what an out-of-order core can extract. The future of computing should surely not rest on terawatts of power burnt to pretend a processor is simpler than it is.</p>
<p>There is some hope in the ideas of upstarts, like <a href="https://millcomputing.com/">Mill Computing</a> and <a href="https://www.tachyum.com/">Tachyum</a>, as well as research ideas like <a href="https://dl.acm.org/citation.cfm?id=3151034">CG-OoO</a>. I don’t know if they will ever find success. I wouldn’t bet on it. Heck, the Mill might never even get far enough to have the opportunity to fail. Yet I find them exciting, and much of the offhand “sounds like Itanium” naysay is uninteresting.</p>
<p>This article focuses on architectures in proportion to how much creative, interesting work they’ve shown in public. This means much of this article comments on the Mill architecture, there is a healthy amount on CG-OoO, and the Tachyum is mentioned only in passing. There are more I could have discussed, like TRIPS (an older dataflow architecture), Denver (NVIDIA’s newish ARM architecture using binary translation) or VISC (Soft Machine’s ‘threadlet’ architecture that got killed by Intel), but I chose to look at the architectures that I consider most promising for general-purpose workloads.</p>
<p>Along with details and comments on these technologies, I have tried to supplement the article with my own untested ideas. These are formulated to get people to focus more directly on the underlying limits of the system, and anyone should feel free to take or extend those ideas if they can make something of them. Not all of my ideas are entirely unique, but I suspect some of them are new.</p>
<p>A typical instruction in a typical instruction set performs something like the following:</p>
<ul>
<li>Read the operands from the global register file.</li>
<li>Perform the operation.</li>
<li>Write the output to the global register file.</li>
</ul>
<p>This looks inoffensive until you start looking at the critical path between instructions, especially with a high level of parallelism and where the operation is something cheap, like addition.</p>
<ul>
<li>Read 16 operands from the global register file simultaneously, at arbitrary locations.</li>
<li>Perform 8 simultaneous operations.</li>
<li>Write the 8 results to the global register file simultaneously, at arbitrary locations.</li>
<li>Read 16 new operands from the global register file simultaneously, and subset of which may be any combination of the just-written results.</li>
<li>Perform 8 simultaneous operations.</li>
<li>…</li>
</ul>
<p>It is difficult enough to figure out how to build a register file that can support this throughput on a cycle-by-cycle basis. It is another to use this as a primary method for routing data between functional units. You can only really mitigate the former issue, but the latter has a standard and inoffensive solution: <a href="https://en.wikipedia.org/wiki/Operand_forwarding">operand forwarding</a>.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*GfQGVmWUWuOerJ2dQAkQsg.png?q=20" height="598"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*GfQGVmWUWuOerJ2dQAkQsg.png" height="598">
</div></div>
<figcaption>If the functional units need their output again at a low latency, they route through the highlighted path. Most such instructions still need to be written to the register file unless they are immediately overwritten, in case another instruction later on requires them.</figcaption></figure><p>This approach helps the situation, but scaling can still be a challenge. A Skylake CPU has <em>348 physical registers</em> per core, split over two different register files. In contrast, there are 8 execution pipelines, only half of which have general ALUs, so the forwarding path is much smaller.</p>
<h2>Coarse-Grained Out-of-Order (CG-OoO)</h2>
<p>Modern cores have enough transistors to build these huge register files — a modern smartphone has billions — but they add latency, long wires and communication costs. The cost also scales with the increasing number of read and write ports. Even more importantly, out-of-order hardware struggles to scale because the core has to keep track of dependencies globally between a vast number of buffered instructions.</p>
<p>CG-OoO is a design intended to break this set of global dependencies whilst still allowing for long-ranged out-of-order execution. The key observation allowing this is that it seems possible to handle dependencies hierarchically, and circumvent the imperfections in the hierarchical approach — inefficient allocation of resources — by overprovisioning the execution resources.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*ixC38jiCUY7Wi26Vn0jRtA.png?q=20" height="269"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*ixC38jiCUY7Wi26Vn0jRtA.png" height="269">
</div></div>
<figcaption>A simplified version of Fig 1. from the <a href="https://dl.acm.org/citation.cfm?id=3151034">CG-OoO paper</a>. Block Windows (BW) feed instructions into execution clusters, each with a number of <em>Execution Units (EU)</em>. A BW contains a small, unshared Local Register File (LRF) and a subset of a shared Global Register File (GRF).</figcaption></figure><p>Each Block Window (BW) in the diagram above is responsible for the execution of a single basic block, a region of code without internal control flow that generally averages 5–20 instructions long. Each Block Window contains two small register files. One is a low-throughput Local Register File (LRF), which is only visible to the executing basic block, and another is a segment of a larger Global Register File (GRF), which communicates globally to all other Block Windows.</p>
<p>The Local Register File is used for approximately 1 in 3 operands and supports a BW with minimal amortized throughput, at only 2 reads and 2 writes per cycle. The LRF is short, but generously provisioned relative to the length of the basic block, 20 entries long for their models, avoiding hazards without renaming. According to the authors, this makes the LRF a factor of 25 cheaper per access than a 256-long heavily-multiported register file.</p>
<p>The Global Register File is shared and requires renaming — that is, the production of a mapping from the source register file indices to GRF indices such that all semanic writes are to different GRF indices, with reuse only after a later write to the source register file index is marked as non-speculative. Renaming maps writes by a Block Window to the local GRF segment, so only register reads are global.</p>
<p>Out-of-order execution also introduces costs for instruction issue and wakeup: In a fully out-of-order processor, any write to the LRF could make any locally-queued instruction ready to issue, and any write the the GRF could make any instruction ready to issue from any BW. Therefore, CG-OoO performs only a small amount of out-of-order execution, hierarchically. Within each BW, only from the first three or so unissued instructions are candidates for dispatch. Distant parallelism is only extracted by parallel execution of BWs.</p>
<p>The corresponding cost reductions are worth listing explicitly.</p>
<ol>
<li>Local instruction reordering is cheap, requiring hazard checking hardware and wakeup only over about three to five local instructions.</li>
<li>Global wakeup affects only those instructions that take global registers and are in the small set of candidate operations for each BW.</li>
<li>Stalled instructions are sufficiently limited in number that they can cache operands inside the issue hardware. They will then be able to read any missing operands in later cycles from the bypass, so only freshly queued instructions ever perform explicit GRF reads.</li>
<li>Each GRF segment requires few ports, since GRF reads will be mostly distributed between segments.</li>
</ol>
<h2>The Mill</h2>
<p>Whereas CG-OoO allows for significant aggregate throughput at the expense of limiting local throughput to a single instruction per cycle amortized, and underperforms when local ILP is unusually large, the Mill first attempts to obliterate local throughput limits and uses secondary methods to extract nonlocal parallelism.</p>
<p>You can get most of the motivation needed to invent the Mill’s belt by noticing how much more headroom operand forwarding has than large global register files, primarily because of the comparatively miniscule number of sources. If the principle operations in a CPU should be those that can be done fast and cheap, the bypass should be the primary source of data movement and multiported register files should only be involved when necessary for long-term storage.</p>
<p>We can prioritize the use of the bypass by making operands more available from it using a small amount of buffering of the outputs at the end of each pipeline. In synergy, we can make register file operations opt-in; the Mill calls reads and writes to their register file substitute <em>fill</em> and <em>spill</em>.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*UTVTSUwQUA4i39qerUWY6g.png?q=20" height="598"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*UTVTSUwQUA4i39qerUWY6g.png" height="598">
</div></div>
<figcaption>The Mill is different yet; their register file is replaced with a ‘scratchpad’, which is allocated per frame and accessed with byte offsets. It has a 3-cycle spill-to-fill latency, and their Silver core can handle only 4 spills and fills per cycle in total — entirely insufficient for a similar-width out-of-order, but easily in excess here.</figcaption></figure><p>With this it’s easy to see that this gives a greater opportunity for the hardware to scale, and detaches register pressure from overall throughput, this raises a few questions.</p>
<ol>
<li>How do you build a bypass that scales with the increased buffering in the system?</li>
<li>How do you avoid this impacting the critical path latency, relative to the smaller, faster bypass in a standard processor?</li>
<li>How much buffering is needed, and how do you prevent this becoming a bottleneck?</li>
<li><strong>How does the program interface with the hardware?</strong></li>
</ol>
<p>The Mill has a straightforward approach to this challenge. A crossbar, a piece of hardware that connects N sources with M sinks, easily scales to the sizes involved here. Unfortunately, larger crossbars have more latency than the original bypass, slowing clock speeds. A solution to this issue is possible by building a fast path into the crossbar.</p>
<p>Latency-1 instructions immediately exit into a designated latency-1 buffer. Higher latency instructions enter corresponding higher-numbered buffers, and latency-1 instructions also migrate to higher-numbered buffers after their first cycle. These latency-1 buffers are small in number and latency-critical, and enter the crossbar fast-path. Instructions with higher latency are less impacted by increased routing delay and often don’t end cleanly on a cycle boundary, so can start their routing earlier in the cycle using the slow-path crossbar.</p>
<p>The amount of buffering needed is of major concern. Luckily, many values are completely temporary, and thus proper scheduling does not need much buffering for these at all. Further, the register file or scratchpad easily handles many longer-lived values. It would seem that few cycles of buffer may well suffice, to cover those outputs that live only just short enough that they cannot be migrated without overhead to the register file or scratchpad.</p>
<p>The Mill resolves some further rough-edges. Operations are migrated between buffers, and automatically spilled to a side buffer, to extend the life of operands in the case of skew pipeline usage. This automated data movement is made practical in part because these pathological cases are those with low throughput, at least in the Mill’s semantic model. When every pipeline is producing values every cycle, operands fall off the belt very fast. The only time operands last a long time on the belt (without explicitly recovered, and ignoring function calls) is when the belt is moving slowly.</p>
<p>The last question about the program-facing interface to the hardware is left open, and touched on in later sections.</p>
<p>Exploring the CG-OoO’s approach shed some light on the trade-offs required in the construction of an out-of-order processor. However, the previous discussion intentionally did not clarify whether the Mill’s approach was for an in-order or out-of-order core. In fact, very little of the Mill’s approach seems to directly improve the cost-benefit ratio of out-of-order hardware, as it would seem all of the complex wakeup logic would still have to exist.</p>
<p>One might be tempted to consider a CG-OoO-Mill hybrid, if just for perspective. However, it is difficult to imagine; the belt is optimized for throughput on values with short lifetimes, so regardless of whether the core supports out-of-order execution, we must confront the challenge of extracting static parallelism directly.</p>
<p>Mill Computing claims that in-order computing is possible to do fast, and they give quite a few tools to get there. Before analyzing these, a causal understanding of the specific benefits out-of-order approach is warranted. The following shortlist, though brief, should be nearly exhaustive.</p>
<ol>
<li>Out-of-order hardware handles false dependencies, such as through condition codes and register hazards.</li>
<li>Out-of-order hardware allows near-optimal scheduling around control flow.</li>
<li>Out-of-order hardware efficiently schedules in the context of variable-latency instructions, particularly for memory access.</li>
</ol>
<h2>1. False dependencies</h2>
<p>Exceptions on integer division by zero and condition codes both create incidental dependencies: exceptions prevent the compiler from reordering instructions outside of loops and ifs, or across globally-visible memory writes, and condition codes make hardware speculation more difficult. The solution to both issues is to put these conditions into software, with division by zero resulting in a register value of some sort, and conditions being handled with booleans and comparisons. This solution is so noncontroversial that even the conservative RISC-V architecture takes this approach. Another issue in this category is exceptions on a failed memory load, but memory is better delegated to its own section due to its complexity.</p>
<p>Register hazards are an unavoidable consequence of using a limited number of names to refer to an unbounded number of source-level values. There are really only two issues from these hazards:</p>
<ol>
<li>Register hazards force storage elements to be used in program order, impeding dynamic reordering without renaming.</li>
<li>Register name limits represent an upper bound on data that can be stored in the primary storage medium without renaming.</li>
</ol>
<p>The Mill is affected by both issues, since the hardware state of the belt varies depending on how the current control block was reached. The Mill uses renaming to solve the first of these issues, and uses a sizeable scratchpad without renaming for the second. Rename hardware is cheap enough that this is not an unacceptable compromise, and the Mill’s rename tables are smaller than a comparable out-of-order processor’s regardless.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*M16sgiZ2ZytAoMH4H2XD9w.png?q=20" height="385"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*M16sgiZ2ZytAoMH4H2XD9w.png" height="385">
</div></div>
<figcaption>Modified Fig 15 (b). from the <a href="https://dl.acm.org/citation.cfm?id=3151034">CG-OoO paper</a>. The energy cost of rename dominates the rest of decode in their modelled processors, with the stage taking 9%, 14% and 0.2% of total execution power for each processor, respectively.</figcaption></figure><p>Whilst rename is an unavoidable cost of this sort of runtime-execution mismatch, it’s far from obvious that full, unstructured rename is the best we can do. In fact, it seems a variant of the CG-OoO architecture might allow for completely removing the rename stage, and the associated ad-hoc global communication.</p>
<p>In an out-of-order CPU, each instruction has its own view of the logical global register file — the program-visible registers at that point in the execution — , which are mapped onto some subset of the physical register file. Because a potentially unique view must exist for all of 200+ instructions in the reorder buffer, it’s important that this state is shared. This is what rename is good at.</p>
<p>In the CG-OoO architecture, each Block Window, not each instruction, has its own view of the logical global register file. Each Block Window writes to any given register at most once, and each basic block has a header that declares which registers will be written to. This makes sharing registers less necessary.</p>
<p>Consider if each Block Window had a complete, separate copy of the logical register file, and propagation happened on write, not read. Then each instruction could use only its local registers. Register forwarding paths could exist separately for each logical register, rather than supporting full, any-register-to-any-register communication.</p>
<p>I can think of two interesting approaches to build this. Each logical register could have an associated bus, and each BW’s copy would be instantiated to know which BW to take its result from. This would be efficient, and would only bottleneck when the same logical register is written to multiple times in the same cycle by different BWs. The second would be to allocate basic blocks to BWs in a specified order, and put each logical register on a special bus that captures the overwrite and not-ready semantics directly.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*Ju5N1NW2jWane11oW05rrw.png?q=20" height="576"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*Ju5N1NW2jWane11oW05rrw.png" height="576">
</div></div>
<figcaption>Basic blocks are allocated to BWs 0, 3, 6, 1, 4, 7, 2, 5, 8, 0, …, in that order, to spread out demand for execution units. Signals can propagate through multiple BWs when the red lines are set, which happens when the corresponding BW will never write to that logical register, so the previous value is preserved. When a value is written, the orange-yellow line is enabled, and the BW’s value is pushed to the bus. The blue lines enable skip connections, which might reduce latency.</figcaption></figure><p>In this approach, each register communicates point-to-point, and there is no contention whatsoever, but the fixed order of BWs limits the overall depth of runahead execution.</p>
<h2>2. Control Flow</h2>
<p>The saving grace of computing has in many ways been the branch predictor. If we assume the predictor is 97% accurate, which is not unreasonable, and branch instructions (both taken and untaken) occur once every 5 instructions, the predictor accurately projects a path on average 5 / 0.03 ≈ 167 instructions into the future. The power of an out-of-order processor is that it is able to reorder instructions as it wills within as much of this window as currently processed. In an out-of-order processor, branch instructions <em>terminate</em> dependency chains — that is, no instruction ever depends or thereby waits on a correctly-predicted branch instruction. So too with CG-OoO, where control flow terminates basic blocks, and many successive basic blocks can evaluate in parallel as their dependencies are resolved.</p>
<p>No in-order approach seems to recognize how much work out-of-order hardware puts in to making control flow a manageable issue — albeit the <em>effort</em> it spends to get that work done is well understood. This is the first junction in this story where I will state I feel the Mill is insufficiently powerful to handle the performance demands of the competition. The frequency of control flow makes getting this right so important as to undermine any approach that fails to handle it well, which makes this discussion particularly interesting.</p>
<p>The approaches to try to extract this kind of behaviour from an in-order machine look at first blush quite limited. Branch prediction is still used to hide the instruction fetch and decode delay, and the Mill has a convincing argument that its predictor could be better than the standard approach, so that should not be a concern, and will not be detailed here. Branch delay slots, where a branch waits a number of cycles before happening, are mostly useless for correctly predicted branches, but the Mill does have a related and interesting idea with their ‘phasing’ concept (albeit they sell this as increasing performance more generally, which it does not). The idea there is equivalent in nature to a branch delay slot a fraction of a cycle long, where sink instructions in the exiting block can interleave with non-sink instructions in the entered block. This ‘tears’ the branch over several cycles, avoiding sparse ‘warm-up’ and ‘cool-down’ instructions where units go underutilized.</p>
<pre><span> standard VLIW phasing (program) phasing (execution)<br>================================================================</span><span> read read modify write read</span><span> read modify read modify write read modify</span><span> read modify write read modify write read modify write<br> =====branch====== ======<br> modify write read modify write read modify write<br> ===branch===<br> write read modify write read modify write<br> =====branch====== =======<br> read read modify write read modify write</span><span> read modify modify write</span><span> read modify write write</span><span> modify write</span><span> write</span></pre>
<p>It should be noted that the delays in the VLIW example is illustrative of a typical dependency chain, but unusual instruction distributions could actually favour a standard VLIW.</p>
<p>The remaining 90% of the problem with control flow is of two parts.</p>
<p><strong>2.1 Branch instructions impose ordering constraints on the program</strong></p>
<p>Let’s break down the first of these concerns with an example.</p>
<pre><span> a = j + k<br> b = x + y<br> c = m * n<br> if c &gt; a<br> / \<br> d = a + 7 d = a - b<br> e = b + d e = d + d</span></pre>
<p>Assume purely for sake of demonstration that addition and subtraction have latency 1, whereas multiplication has latency 2, and that all machines are fully pipelined, have no resource contention, have a superscalar width of 2, and have accurate predictors. Different examples exist for cores with larger scalaraties, but are much harder to diagram.</p>
<p>A standard in-order approach struggles to extract parallelism.</p>
<pre><span> Left branch taken Right branch taken</span><span> c = m * n a = j + k c = m * n a = j + k<br> b = x + y b = x + y<br> c &gt; a c &gt; a<br> d = a + 7 d = a - b<br> e = b + d e = d + d</span></pre>
<p>An out-of-order processor can produce an optimal trace, assuming the compiler is smart enough to prioritize <code>c</code> even without a directly exposed pipeline.</p>
<pre><span> Left branch taken Right branch taken</span><span> c = m * n a = j + k c = m * n a = j + k<br> b = x + y d = a + 7 b = x + y d = a - b<br> e = b + d c &gt; a e = d + d c &gt; a</span></pre>
<p>There are two common directions for mitigating this in-order disadvantage, typically part of what is termed ‘superblock scheduling’ in compilers: moving instructions to after the branch executes (‘hoist below’), or executing them early, before the branch (‘hoist above’). Host below allows you to overlap the critical path of the instructions in one basic block with the critical path of instructions in another, improving scheduling, but it adds a dependency on the branch at runtime. Hoist above removes a dependency on the branch, but requires simultaneous evaluation of both targets <em>even when the core has accurate predictions of the actual branch target.</em></p>
<pre><span> Hoist below<br> Left branch taken Right branch taken</span><span> c = m * n a = j + k c = m * n a = j + k<br> (stall) (stall)<br> c &gt; a c &gt; a<br> b = x + y b = x + y<br> d = a + 7 d = a - b<br> e = b + d e = d + d</span><span> Hoist above<br> Left branch taken Right branch taken</span><span> a = j + k b = x + y a = j + k b = x + y<br> c = m * n d = a + 7 c = m * n d = a + 7<br> D = a - b e = d + d D = a - b e = d + d<br> E = D + D c &gt; a E = D + D c &gt; a</span></pre>
<p>Hoist-above has the additional issue of requiring disambiguation between the two halves of its computation after the fact (you need to pick which values to keep), and risks exponential blowup if you are speculating more than one branch deep. The disambiguation required in hoist-above also forces following instructions to depend on both paths of the computation, independent of whether one side is much shorter and faster than the other. Compromises with less or mixed levels of hoisting suffer a similar mix of these issues.</p>
<p>The Mill intends to tackle this issue mostly with hoist-above, mitigating the cost of these extra computations by having very cheap (sub-cycle) disambiguation instructions and a significantly wider core than its direct competitors. This is wasteful, and still doesn’t schedule instructions efficiently.</p>
<p><strong><em>Idea: Branch Predict and Branch Verify</em></strong></p>
<p><em>Edit: Tapabrata Ghosh (deepnotderp) </em><a href="https://news.ycombinator.com/item?id=19887003"><em>points to</em></a><em> a paper from 2015 that invented this same technique, </em><a href="https://dl.acm.org/citation.cfm?id=2750400"><em>Branch vanguard: decomposing branch functionality into prediction and resolution instructions</em></a><em>.</em></p>
<p>This issue seems to be resolvable in part, with a few key insights. The out-of-order core is able to start executing down the right path ahead of time not because it went out of order, but because it trusted its predictor. An in-order processor can do this too! Let me introduce two very unique instructions.</p>
<p><strong><em>brp (Branch Predict):</em></strong><em> Takes an address to branch to, and takes the branch if and only if the branch predictor suggests to do so. This instruction does not take a predicate as an argument.</em></p>
<p><strong><em>brv (Branch Verify):</em></strong><em> Takes a predicate, and the tag of a </em><strong><em>brp</em></strong><em> to update. If the predicate is false, branches to the other address in the </em><strong><em>brp</em></strong><em> designated by a bit in the instruction; fallthrough if down the taken path, taken if down the fallthrough path. The verify instruction is never predicted to be taken. If the verify is taken, the branch predictor is updated for the corresponding Branch Predict instruction instead.</em></p>
<p>This pair may seem unintuitive at first, but it is really just a simple decomposition of parts of a branch instruction that were already there. This pair of instructions allows getting the good schedules of hoist-below without the dependencies it invokes.</p>
<pre><span> Left branch taken Right branch taken</span><span> brp brp<br> c = m * n a = j + k c = m * n a = j + k<br> b = x + y d = a + 7 b = x + y d = a - b<br> e = b + d brv c ≤ a e = d + d brv c &gt; a</span></pre>
<p>Note that <code>brp</code> exists purely in control-flow, so like static branches it does not take a pipeline slot, and is only displayed to make the order of instructions clear.</p>
<p>There are still struggles with <code>brp</code>, even if one agrees that it improves the situation.</p>
<ul>
<li>In the case of converging control flow, where multiple branches lead to the same block, hoist-below can fail, since incoming branches will want to hoist different instruction sequences. Fortunately, unconditional converging jumps can be tackled well by hoist-above.</li>
<li>Hoist-below duplicates code down both sides of the branch. The cost of code duplication is exponential in the depth.</li>
<li>The target branch must keep around the alternative branch target’s registers, in case of a failed branch verification, which is linear in depth.</li>
<li>All Branch Predict targets become converging targets, including <code>brv</code>, which requires state to be consistent regardless of which parent you come from. <code>brv</code> could handle this with dedicated recovery code or hardware support, whichever works best.</li>
<li>Store instructions cannot be performed speculatively without special hardware support, and load instructions still depend on them.</li>
</ul>
<p>The exponential cost in code duplication, and significant register pressure at higher depths, means <code>brp</code> should still be used fairly shallowly, maybe two deep. This is vastly less than an out-of-order processor can go, so <code>brp</code> is far from a be-all solution on its own.</p>
<p><strong>2.2 Branches outside the current context are opaque to the compiler</strong></p>
<p>A branch instruction to a function, or especially a branch instruction through a pointer, cannot generally have calculations hoisted across it. Static function calls can be inlined, which works pretty well when the cost of inlining is manageable, but those the compiler decides not to, as well as those that cannot be, are a wall to any standard static compiler.</p>
<p>The Mill has some very interesting approaches here, but nothing that changes the overall cost landscape. Function call overhead on the Mill is incredibly low; this is down to some fancy optimizations such as <a href="https://millcomputing.com/topic/control-flow-divergence-and-the-belt/#post-3335">their single-cycle bulk rename</a>. Function calls chain cheaply and don’t interrupt the local instruction schedule. Memory has some interesting but incomplete interactions, which will be covered in a memory section.</p>
<p>Alas, none of this changes the fact that the Mill can’t interleave the evaluation of two called functions. The only directly mitigating factors seem to be that inlining is fairly aggressive, and indirect branches are the most difficult branches for out-of-order hardware too. Perhaps the Mill’s faster calls and lower mispredict penalty will trade equal for this, but the world might not be so kind.</p>
<p>It would be nice to solve this issue, even if it is not problematic to get wrong as the section from before. I do not know how. This concludes the control flow section.</p>
<h2>3. Variable-latency instructions</h2>
<p>Scheduling statically for variable-latency instructions is hard. However, a convenient truth of hardware is that the efficient ways to calculate results are generally timing-agnostic with respect to their inputs. As far as I know, there are only three exceptions in a standard processor that are difficult to avoid.</p>
<ol>
<li>Branches, which are expensive when unpredictable. Hardware handles this by scheduling for the predicted scenario and trying to to avoid mispredicts getting too slow, which works very well.</li>
<li>Division, except by a constant. Division instructions are rare enough not to strictly require efficient handling, with the Mill taking the unusual approach of just doing divides in software (utilizing fast hardware approximations to reduce the cost).</li>
<li><strong>Memory instructions. This is unavoidable and performance critical.</strong></li>
</ol>
<p>Some other, less ‘standard’ instructions exist, such as transcendentals, which I will skip over.</p>
<p>Because the other two concerns are nearly non-issues, only the third point about memory will be addressed in detail. Other potential long-latency instructions can sometimes be modelled as memory accesses. Further, since memory latency interacts with the previously delegated concerns about memory, the topic deserves a section of its own, and this section will be cut short.</p>
<p>If the previous section gave the impression that, say, control flow is a particularly tough problem, but inventive methods can give powerful and practical tools to tackle it… good, that was the intent. In this section I want you to take that expectation and dash it against the rocks — memory is hard<strong><em>.</em></strong> Though there are prospective tools to tackle it, nothing truly looks like a solution, and it’s not clear that the most effective-looking solutions are even practical.</p>
<p>Memory is not just an issue because it is complex — but it is!—; memory is an issue because of how easily it can dwarf the concerns of the rest of the system. Whereas a branch predictor might fail once every hundred and fifty instructions, burning a dozen cycles in its wake each time it does, poorly written code may miss on cache at regular intervals, and a DRAM read takes literally hundreds of cycles. One load instruction can take more time to complete than the combined cost of 500 others, just by missing cache. A single load instruction that causes a TLB miss may take vastly longer still.</p>
<p>An inexhaustive overview of the issues involved will aid the following discussion on how different approaches tackle them.</p>
<ol>
<li>Reordering loads and stores over each other is normally only possible when they do not alias (they refer to different parts of memory). Despite this, hardware frequently wants to start loading from a known address before other addresses that might conflict with it are known.</li>
<li>Memory accesses can fault, which normally results in an exception. This further prevents reordering of memory operations.</li>
<li>Memory accesses take anywhere from a handful of cycles, commonly 3–4 for the top-level cache, to hundreds in the worst case. It is generally impossible provide guaranteed timings needed for static scheduling. Scheduling for throughput generally requires assuming latencies will be short, and scheduling for long latencies requires mixing distant instructions and can have lower throughput since it extends critical paths.</li>
<li>Whether a memory instruction will miss cache is hard to predict. Compilers seldom have any nontrivial understanding of which instructions will be slow.</li>
<li>Any given section of code might have multiple optimal schedules, depending on which instruction miss what levels of cache, if any. The prevalence of memory instructions and the small size of the top-level cache means that this risk is not safe to disregard.</li>
<li>Hardware can handle multiple cache misses simultaneously, termed <em>memory level parallelism</em>. The opportunity to do so goes up when significant cache misses occur, but extracting this parallelism requires the architecture to be able to execute operations correspondingly far ahead in the program during those cache misses.</li>
<li>Speculating on memory can cause side channel attacks, the most famous being Spectre. These concerns will only be discussed at the end of this topic.</li>
</ol>
<h2>Baseline approaches</h2>
<p>Two historical approaches to memory are useful as a baseline to compare against. The simplest approach that significantly improves the situation for in-order processors is to delay a stall on an unfinished memory load until the result is demanded by a later instruction. A load-store queue is used to check that aliasing memory accesses do not reorder improperly relative to each other, and the core stalls to resolve conflicts if necessary. In theory, this approach can even avoid stalls on register-register moves, if proper bookkeeping is done. With appropriate scheduling by the compiler, this technique allows some load latency to be hidden. In certain circumstances a load might even be resolved after branches or function calls.</p>
<p>There are many restrictions to this approach. A load instruction cannot be issued as soon as its address is available if it has the potential to depend on a store later in the program; the hardware semantics mean that the declared program order is canonical. The instruction that first uses (‘<em>forces</em>’) the load is scheduled statically, but the optimal position will depend on whether the load stalls. The load consumes a register as soon as it is issued, which increases the cost of delaying the consumer of the result. A memory operation also cannot be issued speculatively in the case that it might fault. This combination of concerns means the distance between the load and the forcing of the load will normally have to be small, so only small latencies can be hidden by it.</p>
<p><em>Runahead</em> is a much more heavyweight attempt to tackle the problem, with some popularity today, such as in NVIDIA’s Denver architecture. Runahead directs its effort primarily to the problem of extracting memory level parallelism during long memory stalls. After a cache miss has occurred, the processor speculatively executes down the most likely instruction path, ignoring any register values that are not known (eg. the missed load) or take too long to evaluate (eg. other loads that miss cache). When the initial cache miss has resolved, this speculative work is discarded. The intent of this speculative first pass is to start loads early, bringing their values into cache. Although the instructions executed are wasted, after resuming standard execution the core is less likely to miss cache, resulting in overall higher throughput.</p>
<p>Runahead is much cheaper than a traditional out-of-order approach to speculative execution, yet can still comfortably run many hundreds of instructions ahead, through arbitrary control flow and function calls. It tackles the most pathological memory issue, a cold cache, and it can synergize with other approaches. A common extension is to use address prediction, allowing early loads even when addresses depend on register values which are unknown.</p>
<h2>Mill Computing’s approach</h2>
<p>The Mill uses a variant of the first approach to loads, with several key advantages and — (though the company makes no mention of it) — several meaningful disadvantages. Keep in mind that they have also not announced any sort of runahead execution, so their approach needs to handle the entire long-tail distribution of memory stalls.</p>
<p>The specific approach the Mill has changed over time. The initially advertised approach will be discussed here, and the changes they have made will be discussed in a later security section.</p>
<p>Memory reads on the Mill are split into two parts, the initial <code>load</code> that should be scheduled when the address is first generated, and the retire, which occurs either after a fixed number of cycles or on reaching a specific <code>pickup</code> instruction that references the original <code>load</code>. There is also a <code>refuse</code> should the load be unwanted. In order for loads to be scheduled early, over potentially-hazarding stores, waiting loads must check (‘snoop’) outgoing stores. Loads are therefore semantically ordered by the location of their retire, rather than issue. The unit that handles the load and waits for retire is called the retire station. If a memory access faults, a specific ‘NaR’ data value is returned, and no exception is raised.</p>
<p>The fixed-latency <code>load</code>, which produces its value after a fixed cycle count, is appropriate when the consumer is in a known location relative to the producer, which is common since the Mill is statically scheduled. This operation can pass through control flow, but requires all branches to accept a retiring load at the same point, and so is only really intended to be used in branchless code segments or loops. The <code>pickup</code>-based retire is appropriate for diverging control flow, but tags are statically allocated which raises some specific challenges for code with loops or converging control flow that will be discussed later.</p>
<p>Being able to hoist the initial load over other memory operations makes it much easier to schedule early. A load is also allowed to cross a complete function call invocation, such that it retires only after the call is complete. If a called function uses up the limited retire stations (say, 16), load addresses from outer context frames are saved, and reissued upon return, such that the initial <code>load</code> acted only as a prefetch. If the called function does not exhaust the units, the retire station remains active until its natural retire.</p>
<p>Some major issues are visible in this scheme, though some of them take some effort to uncover.</p>
<p><strong>Issue 1: Function boundaries still prevent hoisting loads</strong></p>
<p>Consider the following code:</p>
<pre><span>int f(int x, int *y) {<br> return x + *y;<br>}</span><span>int g(int *x, int *y) {<br> return f(*x, y);<br>}</span></pre>
<p>Unless <code>f</code> is inlined, the Mill is unable to make two important transformations:</p>
<ol>
<li>The load of <code>x</code> inside <code>g</code> must be performed before entry to <code>f</code>.</li>
<li>The load of <code>y</code> inside <code>f</code> cannot be performed until after entry to <code>f</code>.</li>
</ol>
<p>This means that the Mill will never get memory level parallelism from this example. If <code>x</code> and <code>y</code> both miss the cache, you are spending twice the time waiting on memory.</p>
<p>Mill Computing has told me that they do have an approach that tackles this issue which they are not ready to disclose.</p>
<p><strong>Issue 2: Loads potentially in the critical path cannot be delayed</strong></p>
<p>Consider the following code:</p>
<pre><span>inf f(int **x, int **y) {<br> return **x + **y;<br>}</span></pre>
<p>Each of the four load retires are allocated to a statically-determined cycle. The two retires on the left hand side must be in different cycles, and vice-versa for the right. All loads can potentially stall. For a given retire from the left hand side, only one load of the right hand side can potentially be active during that cycle. Therefore a stall in a dereference of <code>x</code> will not always be able to cover a stall in a dereference of <code>y</code>, regardless of how the code is compiled.</p>
<p>This may be illustrated by a specific example. Consider if the first load from each side happens simultaneously, and retire together, and then the second load from each side happens simultaneously, and retire together. If the first load on the left hand side, and the second load from the right hand side misses, those misses cannot be overlapped, despite their proximity in the source code and the availability of the address.</p>
<p>This issue has many forms, not just cascaded loads. Let <code>M</code> represent any expensive computation.</p>
<pre><span>int g(int *x, int *y) {<br> return M(*x) + *M(y);<br>}</span></pre>
<p>If neither load misses, the two computations should mostly overlap. If both loads miss, the two computations should not overlap. No static schedule on the Mill can work for both cases.</p>
<p><strong>Issue 3: Converging dataflow</strong></p>
<p>Consider the following code:</p>
<pre><span>int f(bool cond, int *x, int y) {<br> int ret;<br> if (cond) {<br> ret = *x;<br> } else {<br> ret = 0;<br> }<br> return M(y) + ret;<br>}</span></pre>
<p>The dereference of <code>x</code> would like to be performed in parallel to the evaluation of <code>M(y)</code>. However, only one side of the <code>if</code> performs a load. On the Mill, unifying these branches requires forcing the load, since its value must be available after the branches converge.</p>
<p><strong>Issue 4: It interacts badly with loop pipelining</strong></p>
<p>Loop pipelining is the structuring of a loop to operate over multiple sequential steps of a different iterations of a loop body. Each iteration of the original loop is split across multiple iterations of the pipelined loop, allowing the pipelined loop to operate with the latency of a single <em>stage</em> of the original loop, but the same per-iteration throughput.</p>
<p>The latency-based <code>load</code> retire is intended for this purpose. This fails in two ways. First, and particularly important, the latency of the <code>load</code> is unable to change over the iterations of the loop. Consider the diagram, showing loop iterations phased over multiple cycles.</p>
<pre><span>Optimal schedule when latency is low or loop has few iterations</span><span> read exec write<br> read exec write<br> read exec write<br> read exec write</span><span>Optimal schedule when latency is high or loop has many iterations</span><span> read ... (many cycles) ... exec write<br> read ... (many cycles) ... exec write<br> read ... (many cycles) ... exec write<br> read ... (many cycles) ... exec write</span></pre>
<p>When the Mill schedules like the first example, it is unable to cover very long misses with its loads. When the Mill schedules like the second example, if the loop ended up with few iterations and the load did not miss cache, many cycles would be wasted waiting on a that was available. An out-of-order machine starts off like the first example in early iterations, and transposes into the second example when a cache miss happens, or over many iterations of the loop if not decode limited.</p>
<p>Secondly, any control flow internal to the loop will still prevent a latency-based <code>load</code>. The <code>pickup</code> instruction will have to be used in this case, but it cannot be pipelined over more than a full iteration of the loop because the tags used to force the load are given statically.</p>
<h2>Mill Computing’s approach, cont.</h2>
<p>Mill Computing advertise their approach to memory accesses as roughly competitive to out-of-order hardware. This has been demonstrated to not be the case; not in the small sense like for control flow, where small modifications adjusted things back onto the right path, but in reaching, seemingly absolutist terms. As far as I can tell, the Mill’s approach to memory simply does not seem to come close to fixing memory’s performance gap between in-order and out-of-order hardware.</p>
<h2>Ideas for a better load</h2>
<p>At this point you should have a healthy fear of memory accesses. I will offer four angles to tackle this, but, as previously warned, no approach solves the problem, and some approaches may be expensive or infeasible to implement.</p>
<p><strong>Idea 1: Prodigy-Mill hybrid approach</strong></p>
<p>The Tachyum Prodigy claims very impressive numbers from their simulations; a 4.0 GHz Prodigy is purportedly comparable to a 3.5 GHz Xeon. Their primary disclosed strategy for reducing long-tail memory latency is an unspecified variant of runahead. It is not clear what other techniques were used, but they report a degree of success, outperforming significantly an Itanium core on data cache stall counts.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*jnc-16GZmzRL6vFm58iWqQ.png?q=20" height="569"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*jnc-16GZmzRL6vFm58iWqQ.png" height="569">
</div></div>
<figcaption>source: <a href="https://www.hotchips.org/hc30/2conf/2.09_Tachyum_Tachyum_Hotchips_2018.pdf">https://www.hotchips.org/hc30/2conf/2.09_Tachyum_Tachyum_Hotchips_2018.pdf</a></figcaption></figure><p>Runahead is a very widely applicable technique, and may be responsible for much of the difference between these approaches, especially as the Itanium’s load instructions were not themselves entirely naïve. Runahead should be possible to implement on a Mill-style CPU, which makes it a decent candidate for inclusion.</p>
<p>Mill Computing have told me they have considered the idea and found that unfortunately it did not expose significant opportunities. This contradicts what I’ve read in the literature, so I consider this an interesting idea to follow up on when it is easier to do so.</p>
<p><strong>Idea 2: Restricted Asynchronous Dataflow</strong></p>
<p>One subset of concerns that seems possible to tackle independently is related to chains of multiple loads. This is very common in part to do with the commonly hierarchical structure of objects, where access to data may require chained dereferences. Dereferences like <code>a-&gt;b-&gt;c y</code> or <code>a[i][j]</code> use only a small subset of operations:</p>
<ul>
<li>Memory loads,</li>
<li>Pointer addition with registers or small constants,</li>
<li>Multiplication by small, often power-of-2 constants.</li>
</ul>
<p>The arithmetic here is much simpler than even the most basic arithmetic pipeline. Additionally, loads always have a significant latency; 3 or 4 cycles for top-level cache is common. These make it sound feasible for this case to be handled with dedicated, overprovisioned hardware.</p>
<p>Consider if every retire station output to a specifically-allocated chain of asynchronous ALUs allowing the given set of operations, and that those ALUs were able to be prepared by the main core during the unit’s minimum load latency. Further, allow those ALUs to loop back into the load unit, buffering a chained load, and let this operation happen recursively up to a hardware-defined limit. As an optional extension to the idea, consider allowing the retire station to share any of its produced values with other unit’s ALUs. That allows operations like <code>a-&gt;b-&gt;c + a-&gt;b-&gt;d</code> to happen fully asynchronously.</p>
<figure><div><div>
<div><img width="700" src="https://miro.medium.com/max/60/1*yqlHPqt35PoKkWyB_ft0lw.png?q=20" height="520"></div>
<img width="700" src="https://miro.medium.com/max/1400/1*yqlHPqt35PoKkWyB_ft0lw.png" height="520">
</div></div>
<figcaption>The path back to the address is for chained loads, and latency-critical. The dotted path is for communication between load units, requires some kind of crossbar, and is unlikely to be latency critical. The ALUs are asynchronous and buffer their sideloaded inputs. Buffering in the ALUs and output register is designed to handle multiply-dereferenced values.</figcaption></figure><p>The arithmetic capabilities of this design are naturally limited, and even with the optional inter-unit extension the communication between units is very restricted. As a design for a processor, this would be hopeless. However, it seems plausible that that this could be a realizable concept for the restricted cases given, which opens the door for more asynchronous memory-level parallelism without fully out-of-order hardware.</p>
<p>Two design decisions not discussed are how to enforce memory ordering with potentially-conflicting stores, and how the crossbar is controlled.</p>
<p><strong>Idea 3: Source-agnostic argument pickups</strong></p>
<p>Likely the first thing stopping further delay of a <code>pickup</code> instruction is its inability to handle function calls or converging control flow in the general case.</p>
<p>The first baseline approach, performing <code>pickup</code> implicitly on first use, avoids this problem but means the only known serialization point is the initial <code>load</code>. A simple hybrid approach is possible. A <code>preload</code> instruction replaces the current <code>load</code> and produces a tag that resides on the belt, referring to a retire unit in a <code>prefetch</code> state. A <code>load</code> instruction takes a tag to a <code>preload</code> state retire unit and changes it into a <code>load</code> state, which is the same as <code>preload</code> except that the hardware guarantees order of <code>load</code>s is conserved. A <code>pickup</code> instruction takes <em>any</em> belt value; if it is a value, that value is left unchanged, if it is a tag then the corresponding load is forced and the tag is replaced <em>in place</em> with the value.</p>
<p>If functions must <code>pickup</code> all of their arguments before use (with instructions to make that space-efficient), or have a way to signal to the caller which instructions will be picked up, then loads can cross function boundaries. Because the tags reside on the belt at all times, and <code>pickup</code> is idempotent (you don’t need to know if the value requires <code>pickup</code>), they can handle more arbitrary control flow and pipelining than the standard approach.</p>
<p><strong>Idea 4: Standard prefetch</strong></p>
<p>One of the issues mentioned before was around long-latency instructions in pipelined loops. The solution here is surprisingly simple. The paper <a href="https://www.cl.cam.ac.uk/~sa614/papers/Software-Prefetching-CGO2017.pdf">Software Prefetching for Indirect Memory Accesses</a> shows how to use prefetch instructions for pipelinable loops. This should be considered mandatory for in-order hardware.</p>
<p>A major concern with prefetch instructions is that many prefetches are incorrect or unnecessary, wasting hardware bandwidth. For prefetches that conflict with automatic hardware prefetching, compiler heuristics should suffice. Other prefetches require more finesse. Prefetching 32 loads ahead in a loop is effective if the loop will last 32 load instructions more, but otherwise the prefetch is wasted memory bandwidth. Hardware heuristics for which prefetches are useful, including perhaps a prefetch warmup phase for loops, is an approach that would follow historic trends. Another possibility is a conditional prefetch instruction, allowing efficient suppression of prefetches that are known to go off the end of a loop.</p>
<h2>Security</h2>
<p>Computers are systematically vulnerable to side-channel attacks. For example, an operation may take more power with some inputs than others; therefore information about the value is learnt by looking at power usage. The most common and dangerous attacks are timing attacks. Control flow is one of the largest contributors. The largest, memory, is the topic of this section.</p>
<p>Whenever an address <code>a</code> is used to access memory, the state of the cache may be changed in a way that allows side-channel attacks to discover <code>a</code>. Therefore all values used as addresses should be considered leaked.</p>
<p>The typical response to this has been dismissal — <em>deal with it</em>—because side channel attacks are incredibly difficult to mitigate or avoid, and secure keys and encryption could generally manually circumvent these issues with careful coding. Spectre is a particular instantiation of this exploit that can read arbitrary memory from a process in spite of these safeguards, which is why it is considered such a big deal.</p>
<p>A twice-indirect memory access of the form <code>**a</code> leaks <code>*a</code>. It is common for <code>a</code> to be untrusted and restricted to a given range of allowed values. However, if the values to be accessed are guarded only by control flow, speculative execution may still start a load of <code>*a</code> early, leaking data chosen by an adversary.</p>
<p>This is a major problem, invalidating almost all of the approaches mentioned so far in their naïve incarnations. This affects every out-of-order processor I know of, and the mitigations for it are largely incomplete and hacky. Since CG-OoO is just an out-of-order processor from a memory perspective, that too is vulnerable.</p>
<p>More surprisingly, the Mill’s approach is also susceptible to this. Their <code>load</code> instructions are designed to be hoisted before control flow, and efficient compilation exposes the same side channel as seen in out-of-order processors. <a href="https://millcomputing.com/blog/wp-content/uploads/2018/01/Spectre.03.pdf">Mill Computing published a response to the discovery</a>, claiming that “the Mill architecture is fundamentally immune to Spectre-like vulnerabilities” but instead that their vulnerability to Spectre was “a software bug in the Mill tool chain”. For the first time known to me, the conditional load instructions, <code>loadtr</code> and <code>loadfl</code>, are introduced, which allow for non-speculative condition checking.</p>
<p>This response is concerning. Ultimately, no manufacturer of out-of-order cores has responded to Spectre by abandoning speculation. Early mitigations did so at a heavy performance penalty; future processors will solve the hardest issues in hardware. It seems that either you solve this problem in hardware or you become uncompetitive; the Mill seems to be walking the latter path.</p>
<p>There are several approaches that hardware has which seem promising in avoiding Spectre. For instance, one can attempt to make loads delay writes to cache, allowing side-effect free speculation recovery. One can also focus specifically on the known exploitable variants, and prevent only that subset of speculation that leaks attacker-chosen data, such as by preventing the value of loads to be used speculatively, even if the instruction can be issued speculatively.</p>
<p>However, regardless of whether a mitigation exists, fast memory access is so vital to competitive performance that flushing state on context-switch would be a still be a better compromise than avoiding heavy speculation. This is an issue that must be prioritized to a major degree.</p>
<p>Though this article has ended on a negative note, I do not want that to be the main take-away from this post. Rather, consider that the space of possibilities that is available to us is vastly larger than the space of things that has been tried. These architectures might not be perfect, but they’re making real steps towards solving difficult problems. I have given my own creative ideas in this post, but by no means do CG-OoO, the Mill, and my ideas cover every possible choice available.</p>
<p>If there is some more general thing to be taken from this post, perhaps the following would make a good summary.</p>
<h2>A. Throughput</h2>
<ol>
<li>The architecture must be able to route instructions efficiently and with minimal latency between a large number of functional units.</li>
<li>The architecture should be designed to handle decode and dispatch in significant excess of the available arithmetic ILP.</li>
</ol>
<h2>B. Dependencies and ILP extraction</h2>
<ol>
<li>The architecture must avoid creating false dependencies of any kind.</li>
<li>The architecture should have a wide receptive field over the source-level program in order to see the full breadth of executable instructions.</li>
</ol>
<p>This is a big ask, but both CG-OoO and the Mill make good steps towards them, and the future, though mysterious and uncertain, holds a good deal of promise.</p>
</div></section></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium https://medium.com<br>
<br>
August 1, 2019 at 03:54PM
